File: DD_python/main_knapsack.py
import os

from SourceCode.DD import DD
from Examples.KnapsackInstance.KnapsackInstance import KnapsackStructure
from Examples.KnapsackInstance.KnapsackProblem import KnapsackProblem
from Examples.KnapsackInstance.KnapsackGurobiClass import KnapsackGurobi
from Examples.GuiStructure import CutType
from SourceCode.GraphAlgorithms.ShortestLongestPath.ShortestLongestPath import ShortestLongestPath

## ============================
## SETUP PROBLEM AND PARAMETERS
## ============================

# Code Parameters
_verbose = False
_width = 3

# Instance file
source_directory: str = os.path.dirname(os.path.abspath(__file__))
filename: str = "../DataInstances/Knapsack/Custom/knapsack_instance_v5_d10_seed1.txt"
file_path: str = f"{source_directory}/{filename}"

# Load instance
knapsack_instance = KnapsackStructure(file_path)
print("Instance loaded:")
print(knapsack_instance)

# Build KnapsackProblem for DD construction
knapsack_problem = KnapsackProblem(knapsack_instance)

## ============================
## CREATE EXACT DD
## ============================

dd_exact = DD(knapsack_problem)
dd_exact.create_decision_diagram(verbose=_verbose)
dd_exact.reduce_decision_diagram(verbose=_verbose)

print("\n-- Exact DD --")
print(f"\tConstruction time (sec): {dd_exact.get_building_time()}")
print(f"\tReduction time   (sec): {dd_exact.get_reduction_time()}")
print(f"\tNodes: {dd_exact.get_decision_diagram().get_node_count()}")
print(f"\tArcs:  {dd_exact.get_decision_diagram().get_arc_count()}")

## ============================
## CREATE RELAXED DD
## ============================

dd_relaxed = DD(knapsack_problem)
dd_relaxed.create_relax_priority_decision_diagram(max_width=_width, verbose=_verbose)
dd_relaxed.reduce_decision_diagram(verbose=_verbose)

print("\n-- Relaxed DD (width={}) --".format(_width))
print(f"\tConstruction time (sec): {dd_relaxed.get_building_time()}")
print(f"\tReduction time   (sec): {dd_relaxed.get_reduction_time()}")
print(f"\tNodes: {dd_relaxed.get_decision_diagram().get_node_count()}")
print(f"\tArcs:  {dd_relaxed.get_decision_diagram().get_arc_count()}")

## ============================
## CREATE RESTRICTED DD
## ============================

dd_restricted = DD(knapsack_problem)
dd_restricted.create_restricted_decision_diagram(max_width=_width, verbose=_verbose)
dd_restricted.reduce_decision_diagram(verbose=_verbose)

print("\n-- Restricted DD (width={}) --".format(_width))
print(f"\tConstruction time (sec): {dd_restricted.get_building_time()}")
print(f"\tReduction time   (sec): {dd_restricted.get_reduction_time()}")
print(f"\tNodes: {dd_restricted.get_decision_diagram().get_node_count()}")
print(f"\tArcs:  {dd_restricted.get_decision_diagram().get_arc_count()}")

# Export graph in .gml format
dd_exact.export_graph_file("knapsack_exact_dd_file")
dd_relaxed.export_graph_file("knapsack_relaxed_dd_file")
dd_restricted.export_graph_file("knapsack_restricted_dd_file")

## ============================
## SHORTEST / LONGEST PATH
## ============================

# Optimal solution via exact DD
longest_path = ShortestLongestPath(dd_exact)
longest_path.set_parameters(knapsack_instance.objective_weights, "max")
answer = longest_path.solve()

print("\n-- Optimal solution (exact DD longest path) --")
print(f"\tOptimal value:  {answer.value}")
print(f"\tPath:           {answer.path_print}")
print(f"\tSolution time:  {longest_path.get_time()} sec")

# Dual (upper) bound via relaxed DD
longest_path_relaxed = ShortestLongestPath(dd_relaxed)
longest_path_relaxed.set_parameters(knapsack_instance.objective_weights, "max")
answer_relaxed = longest_path_relaxed.solve()

print("\n-- Dual bound (relaxed DD longest path) --")
print(f"\tUpper bound:   {answer_relaxed.value}")
print(f"\tSolution time: {longest_path_relaxed.get_time()} sec")

# Primal (lower) bound via restricted DD
longest_path_restricted = ShortestLongestPath(dd_restricted)
longest_path_restricted.set_parameters(knapsack_instance.objective_weights, "max")
answer_restricted = longest_path_restricted.solve()

print("\n-- Primal bound (restricted DD longest path) --")
print(f"\tLower bound:   {answer_restricted.value}")
print(f"\tSolution time: {longest_path_restricted.get_time()} sec")

## ============================
## GUROBI – BASELINE
## ============================

print("\n============================")
print("Gurobi baseline (no DD cuts)")
print("============================")
gurobi_base = KnapsackGurobi(knapsack_instance, continuous=False)
gurobi_base.create_model()
gurobi_base.optimize_model()
print(f"Optimization time: {gurobi_base.get_optimization_time()} sec")

## ============================
## GUROBI – DD CUTS
## ============================

print("\n============================")
print("Gurobi with DD cuts")
print("============================")
gurobi_cuts = KnapsackGurobi(knapsack_instance, continuous=False)
gurobi_cuts.create_model()
gurobi_cuts.optimize_with_cuts([dd_relaxed], cut_type=CutType.JOINT_FLOW, cut_strengthening=False)
print(f"Optimization time: {gurobi_cuts.get_optimization_time()} sec")

File: DD_c_plusplus/main_knapsack.cpp
#include <iostream>
#include <string>
#include <vector>
#include <filesystem>

#include "SourceCode/DD.h"
#include "SourceCode/GraphAlgorithms/ShortestLongestPath/ShortestLongestPath.h"
#include "Examples/KnapsackInstance/KnapsackInstance.h"
#include "Examples/KnapsackInstance/KnapsackProblem.h"
#include "Examples/KnapsackInstance/KnapsackGurobiClass.h"

using namespace std;
namespace fs = filesystem;

int main() {

    // ============================
    // SETUP PROBLEM AND PARAMETERS
    // ============================

    bool verbose = false;
    int  width   = 3;

    // Instance file (same as Python main_knapsack.py)
    string source_directory = fs::current_path().parent_path().string();
    string file_path = source_directory + "/DataInstances/Knapsack/Custom/knapsack_instance_v5_d10_seed1.txt";

    // Load instance
    KnapsackInstance knapsack_instance(file_path);

    // Build KnapsackProblem for DD construction
    KnapsackProblem knapsack_problem(knapsack_instance);

    // ============================
    // CREATE EXACT DD
    // ============================

    DD<int> dd_exact(knapsack_problem);
    dd_exact.create_decision_diagram(verbose);
    dd_exact.reduce_decision_diagram(verbose);

    cout << "\n-- Exact DD --\n";
    cout << "\tConstruction time (sec): " << dd_exact.get_building_time()                     << "\n";
    cout << "\tReduction time   (sec): "  << dd_exact.get_reduction_time()                    << "\n";
    cout << "\tNodes: "                   << dd_exact.get_decision_diagram()->get_node_count() << "\n";
    cout << "\tArcs:  "                   << dd_exact.get_decision_diagram()->get_arc_count()  << "\n";

    // ============================
    // CREATE RELAXED DD
    // ============================

    DD<int> dd_relaxed(knapsack_problem);
    dd_relaxed.create_relax_priority_decision_diagram(width, verbose);
    dd_relaxed.reduce_decision_diagram(verbose);

    cout << "\n-- Relaxed DD (width=" << width << ") --\n";
    cout << "\tConstruction time (sec): " << dd_relaxed.get_building_time()                     << "\n";
    cout << "\tReduction time   (sec): "  << dd_relaxed.get_reduction_time()                    << "\n";
    cout << "\tNodes: "                   << dd_relaxed.get_decision_diagram()->get_node_count() << "\n";
    cout << "\tArcs:  "                   << dd_relaxed.get_decision_diagram()->get_arc_count()  << "\n";

    // ============================
    // CREATE RESTRICTED DD
    // ============================

    DD<int> dd_restricted(knapsack_problem);
    dd_restricted.create_restricted_decision_diagram(width, verbose);
    dd_restricted.reduce_decision_diagram(verbose);

    cout << "\n-- Restricted DD (width=" << width << ") --\n";
    cout << "\tConstruction time (sec): " << dd_restricted.get_building_time()                     << "\n";
    cout << "\tReduction time   (sec): "  << dd_restricted.get_reduction_time()                    << "\n";
    cout << "\tNodes: "                   << dd_restricted.get_decision_diagram()->get_node_count() << "\n";
    cout << "\tArcs:  "                   << dd_restricted.get_decision_diagram()->get_arc_count()  << "\n";

    // Export graph in .gml format
    dd_exact.export_graph_file("knapsack_exact_dd_file");
    dd_relaxed.export_graph_file("knapsack_relaxed_dd_file");
    dd_restricted.export_graph_file("knapsack_restricted_dd_file");

    // ============================
    // SHORTEST / LONGEST PATH
    // ============================

    // Optimal solution via exact DD
    ShortestLongestPath longest_path_exact(dd_exact);
    longest_path_exact.set_parameters(knapsack_instance.objective_weights, "max");
    auto answer = longest_path_exact.solve();
    cout << "\n-- Optimal solution (exact DD longest path) --\n";
    cout << "\tOptimal value:  " << answer.value      << "\n";
    cout << "\tPath:           " << answer.path_print << "\n";
    cout << "\tSolution time:  " << longest_path_exact.get_time() << " sec\n";

    // Dual (upper) bound via relaxed DD
    ShortestLongestPath longest_path_relaxed(dd_relaxed);
    longest_path_relaxed.set_parameters(knapsack_instance.objective_weights, "max");
    answer = longest_path_relaxed.solve();
    cout << "\n-- Dual bound (relaxed DD longest path) --\n";
    cout << "\tUpper bound:   " << answer.value << "\n";
    cout << "\tSolution time: " << longest_path_relaxed.get_time() << " sec\n";

    // Primal (lower) bound via restricted DD
    ShortestLongestPath longest_path_restricted(dd_restricted);
    longest_path_restricted.set_parameters(knapsack_instance.objective_weights, "max");
    answer = longest_path_restricted.solve();
    cout << "\n-- Primal bound (restricted DD longest path) --\n";
    cout << "\tLower bound:   " << answer.value << "\n";
    cout << "\tSolution time: " << longest_path_restricted.get_time() << " sec\n";

    // ============================
    // GUROBI – BASELINE
    // ============================

    cout << "\n============================\n";
    cout << "Gurobi baseline (no DD cuts)\n";
    cout << "============================\n";
    KnapsackGurobiClass gurobi_base(knapsack_instance, false);
    gurobi_base.create_model();
    gurobi_base.optimize_model();
    cout << "Optimization time: " << gurobi_base.get_optimization_time() << " sec\n";

    // ============================
    // GUROBI – DD CUTS
    // ============================

    cout << "\n============================\n";
    cout << "Gurobi with DD cuts\n";
    cout << "============================\n";
    KnapsackGurobiClass gurobi_cuts(knapsack_instance, false);
    gurobi_cuts.create_model();
    gurobi_cuts.optimize_with_cuts(vector<DD<int>*>{&dd_relaxed}, CutType::JointFlow, /*cut_strengthening=*/false);
    cout << "Optimization time: " << gurobi_cuts.get_optimization_time() << " sec\n";

    return 0;

}
